core/net/link-stats.c

```diff
@@ -0,0 +1,212 @@
+uint16_t
+guess_etx_from_rssi(const struct link_stats *stats)
+{
+  if(stats != NULL) {
+    if(stats->rssi == 0) {
+      return ETX_INIT * ETX_DIVISOR;
+    } else {
+      /* A rough estimate of PRR from RSSI, as a linear function where:
+       *      RSSI >= -60 results in PRR of 1
+       *      RSSI <= -90 results in PRR of 0
+       * prr = (bounded_rssi - RSSI_LOW) / (RSSI_DIFF)
+       * etx = ETX_DIVOSOR / ((bounded_rssi - RSSI_LOW) / RSSI_DIFF)
+       * etx = (RSSI_DIFF * ETX_DIVOSOR) / (bounded_rssi - RSSI_LOW)
+       * */
+#define ETX_INIT_MAX 3
+#define RSSI_HIGH -60
+#define RSSI_LOW  -90
+#define RSSI_DIFF (RSSI_HIGH - RSSI_LOW)
+      uint16_t etx;
+      int16_t bounded_rssi = stats->rssi;
+      bounded_rssi = MIN(bounded_rssi, RSSI_HIGH);
+      bounded_rssi = MAX(bounded_rssi, RSSI_LOW + 1);
+      etx = RSSI_DIFF * ETX_DIVISOR / (bounded_rssi - RSSI_LOW);
+      return MIN(etx, ETX_INIT_MAX * ETX_DIVISOR);
+    }
+  }
+  return 0xffff;
+}
+/*---------------------------------------------------------------------------*/
+/* Packet sent callback. Updates stats for transmissions to lladdr */
+void
+link_stats_packet_sent(const linkaddr_t *lladdr, int status, int numtx)
+{
+  struct link_stats *stats;
+  uint16_t packet_etx;
+  uint8_t ewma_alpha;
+
+  if(status != MAC_TX_OK && status != MAC_TX_NOACK) {
+    /* Do not penalize the ETX when collisions or transmission errors occur. */
+    return;
+  }
+
+  stats = nbr_table_get_from_lladdr(link_stats, lladdr);
+  if(stats == NULL) {
+    /* Add the neighbor */
+    stats = nbr_table_add_lladdr(link_stats, lladdr);
+    if(stats != NULL) {
+      stats->etx = LINK_STATS_INIT_ETX(stats);
+    } else {
+      return; /* No space left, return */
+    }
+  }
+
+  /* Update last timestamp and freshness */
+  stats->last_tx_time = clock_time();
+  stats->freshness = MIN(stats->freshness + numtx, FRESHNESS_MAX);
+
+  /* ETX used for this update */
+  packet_etx = ((status == MAC_TX_NOACK) ? ETX_NOACK_PENALTY : numtx) * ETX_DIVISOR;
+  /* ETX alpha used for this update */
+  ewma_alpha = link_stats_is_fresh(stats) ? EWMA_ALPHA : EWMA_BOOTSTRAP_ALPHA;
+
+  /* Compute EWMA and update ETX */
+  stats->etx = ((uint32_t)stats->etx * (EWMA_SCALE - ewma_alpha) +
+      (uint32_t)packet_etx * ewma_alpha) / EWMA_SCALE;
+}
+/*---------------------------------------------------------------------------*/
+/* Packet input callback. Updates statistics for receptions on a given link */
+void
+link_stats_input_callback(const linkaddr_t *lladdr)
+{
+  struct link_stats *stats;
+  int16_t packet_rssi = packetbuf_attr(PACKETBUF_ATTR_RSSI);
+
+  stats = nbr_table_get_from_lladdr(link_stats, lladdr);
+  if(stats == NULL) {
+    /* Add the neighbor */
+    stats = nbr_table_add_lladdr(link_stats, lladdr);
+    if(stats != NULL) {
+      /* Initialize */
+      stats->rssi = packet_rssi;
+      stats->etx = LINK_STATS_INIT_ETX(stats);
+    } else {
```

Comment:
I think that you should `return`whether you manage to add the neighbor or not. If you don't return after `stats->etx = LINK_STATS_INIT_ETX(stats);`, then you will execute the RSSI EWMA even if the neighbor was just added. In that case, the EWMA does not change the value of `stats->rssi` because  `stats->rssi` and `packet_rsssi`are equal. Then, the updated function would look like:
```
void
link_stats_input_callback(const linkaddr_t *lladdr)
{
  struct link_stats *stats;
  int16_t packet_rssi = packetbuf_attr(PACKETBUF_ATTR_RSSI);

  stats = nbr_table_get_from_lladdr(link_stats, lladdr);
  if(stats == NULL) {
    /* Add the neighbor */
    stats = nbr_table_add_lladdr(link_stats, lladdr, NBR_TABLE_REASON_LINK_STATS, NULL);
    if(stats != NULL) {
      /* Initialize */
      stats->rssi = packet_rssi;
      stats->etx = LINK_STATS_INIT_ETX(stats);
    }
    return;
  }

  /* Update RSSI EWMA */
  stats->rssi = ((int32_t)stats->rssi * (EWMA_SCALE - EWMA_ALPHA) +
      (int32_t)packet_rssi * EWMA_ALPHA) / EWMA_SCALE;
}
```